osu.Game.Rulesets.Catch/Edit/CatchSelectionHandler.cs

```diff
@@ -76,21 +76,38 @@ public override bool HandleFlip(Direction direction, bool flipOverOrigin)
 
         public override bool HandleReverse()
         {
+            var hitObjects = EditorBeatmap.SelectedHitObjects;
+
             double selectionStartTime = SelectedItems.Min(h => h.StartTime);
             double selectionEndTime = SelectedItems.Max(h => h.GetEndTime());
 
-            EditorBeatmap.PerformOnSelection(hitObject =>
+            var newComboOrder = hitObjects
+                .OfType<CatchHitObject>()
+                .Select(obj => obj.NewCombo)
+                .ToList();
+
+            newComboOrder.Reverse();
+
+            foreach (var h in hitObjects)
             {
-                hitObject.StartTime = selectionEndTime - (hitObject.GetEndTime() - selectionStartTime);
+                h.StartTime = selectionEndTime - (h.GetEndTime() - selectionStartTime);
 
-                if (hitObject is JuiceStream juiceStream)
+                if (h is JuiceStream juiceStream)
                 {
                     juiceStream.Path.Reverse(out Vector2 positionalOffset);
                     juiceStream.OriginalX += positionalOffset.X;
                     juiceStream.LegacyConvertedY += positionalOffset.Y;
                     EditorBeatmap.Update(juiceStream);
                 }
-            });
+            }
+
+            int i = 0;
+            foreach (bool newCombo in newComboOrder)
+            {
+                hitObjects.OfType<CatchHitObject>().ToList()[i].NewCombo = newCombo;
```

Comment:
This is a bit :exploding_head: 

- is there even a guarantee that `hitObjects` is going to match `newComboOrder` order-wise? i.e. are there any sorting guarantees on `SelectedHitObjects`? not sure
- `.ToList()` in a loop just to get `i`-th element - enumerate once at the start, or if you _really_ don't want to for whatever reason, `ElementAt()` exists

Reply:
> is there even a guarantee that hitObjects is going to match newComboOrder order-wise? i.e. are there any sorting guarantees on SelectedHitObjects? not sure

I only noticed this now, when selecting 1 > 3 then 2. To solve this i added `.OrderBy(obj => obj.StartTime)` to hitObjects so no matter how the user selects them they are always the same.



Also addressed your second point, here is how the method is looking now (same applies to catch):

```C#
        public override bool HandleReverse()
        {
            var hitObjects = EditorBeatmap.SelectedHitObjects
                .OfType<OsuHitObject>()
                .OrderBy(obj => obj.StartTime);

            double endTime = hitObjects.Max(h => h.GetEndTime());
            double startTime = hitObjects.Min(h => h.StartTime);

            bool moreThanOneObject = hitObjects.Count() > 1;

            var newComboOrder = hitObjects.Select(obj => obj.NewCombo).ToList();

            foreach (var h in hitObjects)
            {
                if (moreThanOneObject)
                    h.StartTime = endTime - (h.GetEndTime() - startTime);

                if (h is Slider slider)
                {
                    slider.Path.Reverse(out Vector2 offset);
                    slider.Position += offset;
                }
            }

            int i = 0;
            foreach (bool newCombo in newComboOrder)
            {
                hitObjects.ElementAt(i).NewCombo = newCombo;
                i++;
            }

            return true;
        }
```

Reply:
Move the `.ToList()` call to the front, i.e. at the end of assignment to `hitObjects`. Then you can use `Count` rather than linq `Count()` and the indexer rather than linq `ElementAt()`.

Reply:
This somehow broke it, i've added some logs to see what happened by what they say it should be working fine
but it is assigning a new combo to the end of the selection even tho the logs explicitly say: `False - 3`
```
[runtime] 2024-03-26 10:54:15 [verbose]: 165
[runtime] 2024-03-26 10:54:15 [verbose]: 396
[runtime] 2024-03-26 10:54:15 [verbose]: 626
[runtime] 2024-03-26 10:54:15 [verbose]: 857.3076923076931
[runtime] 2024-03-26 10:54:15 [verbose]: True - 0
[runtime] 2024-03-26 10:54:15 [verbose]: False - 1
[runtime] 2024-03-26 10:54:15 [verbose]: False - 2
[runtime] 2024-03-26 10:54:15 [verbose]: False - 3
```
![image](https://github.com/ppy/osu/assets/90941580/b50bb0e5-9d55-4423-b8a6-fc73759ff6bb)

```C#
        public override bool HandleReverse()
        {
            var hitObjects = EditorBeatmap.SelectedHitObjects
                .OfType<OsuHitObject>()
                .OrderBy(obj => obj.StartTime)
                .ToList();

            double endTime = hitObjects.Max(h => h.GetEndTime());
            double startTime = hitObjects.Min(h => h.StartTime);

            bool moreThanOneObject = hitObjects.Count > 1;

            var newComboOrder = hitObjects.Select(obj => obj.NewCombo);

            foreach (var h in hitObjects)
            {
                Logger.Log($"{h.StartTime}");
                if (moreThanOneObject)
                    h.StartTime = endTime - (h.GetEndTime() - startTime);

                if (h is Slider slider)
                {
                    slider.Path.Reverse(out Vector2 offset);
                    slider.Position += offset;
                }
            }

            int i = 0;
            foreach (bool newCombo in newComboOrder)
            {
                Logger.Log($"{newCombo} - {i}");
                hitObjects[i].NewCombo = newCombo;
                i++;
            }

            return true;
        }
```

Reply:
You were implicitly relying on multiple enumration of `hitObjects` there. Sort the items again by start time after reversing and before restoring the combo order.

Reply:
Sorting again was also needed but the actual issue was at line `hitObjects[i].NewCombo = newCombo;`. This line was "tainting" with 'newComboOrder', after the first iteration of the loop it altered the 'hitObjects' variable making 'newComboOrder' update as well, and index 0 of 'hitObjects' is the last index of newComboOrder, so after the first iteration 'newComboOrder' goes from [True, False, False] to [True, False, True]. TLDR the fix is to have 'newComboOrder' be independent of 'hitObjects', which i did by  changing it to `var newComboOrder = hitObjects.Select(obj => obj.NewCombo).ToList();` and thats just reverting what you recommended me to do so im not sure how to proceed

Reply:
> which i did by changing it to var newComboOrder = hitObjects.Select(obj => obj.NewCombo).ToList(); and thats just reverting what you recommended me to do so im not sure how to proceed

It's kinda not. Fine to do that.

Reply:
Ok so i ll do that